### PR TITLE
Write hls unit tests for get references

### PR DESCRIPTION
**Pull request by peterwicksstringfield**
https://github.com/haskell/haskell-language-server/pull/704

### CODE REVIEW

**Comment by wz1000** [comment]
Very nice, this looks quite helpful.

I think we will need a few test messages to indicate that indexing has completed and synchronize to avoid race condition style failures in tests(i.e. we ask for references too early and get an incomplete result), so this is what I have been working on. This might be the cause of some of the failures you are seeing.

Sorry for the late response, github didn't notify me for either this PR or the comment on the draft HLS PR for some reason. It's very late tonight, so I will take a proper look tomorrow.

**Comment by peterwicksstringfield** [comment]
I figured you were busy, it's not a problem. :) Dunno why it didn't notify you.

The nix builds are failing with something dependency related:

```
building '/nix/store/bnb2xhjh7dz1pgsld2rd3l60mxx8k8h2-cabal2nix-haskell-language-server.drv'...
installing
error: anonymous function at /nix/store/lh6n286kd27aczmqi6jcik37v0ssc91n-cabal2nix-haskell-language-server/default.nix:1:1 called without required argument 'hiedb', at /nix/store/51gm3gjl7i2n3f7pamrbsckm4h0fmhy2-nixpkgs-src/pkgs/development/haskell-modules/make-package-set.nix:87:27
(use '--show-trace' to show detailed location information)
Error: Process completed with exit code 1.
```

And the pre 8.10 builds are failing like this:
```
src/Development/IDE/Spans/AtPoint.hs:260:13: error:
    • The constructor ‘FunTy’ should have 2 arguments, but has been given 3
    • In the pattern: FunTy _ a b
      In a case alternative: FunTy _ a b -> getTypes [a, b]
      In the second argument of ‘($)’, namely
        ‘\case
           TyVarTy n -> [Var.varName n]
           AppTy a b -> getTypes [a, b]
           TyConApp tc ts -> tyConName tc : getTypes ts
           ForAllTy _ t -> getTypes [t]
           FunTy _ a b -> getTypes [a, b]
           CastTy t _ -> getTypes [t]
           _ -> []’
    |
260 |             FunTy _ a b -> getTypes [a,b]
    |             ^^^^^^^^^^^
cabal: Failed to build ghcide-0.6.0.2 (which is required by bench:benchHist
from ghcide-0.6.0.2).
```

So it's not race condition related, the build is simply broken.

Note the review comment below though.

File: test/functional/Reference.hs [hunk 1]
```diff
@@ -2,34 +2,191 @@ module Reference (tests) where
...
+    -- TODO: Something needs to change ...
+    -- These tests take forever anyway while HLS does stuff with cabal.
+    _ <- openDoc "exe/Main.hs" "haskell"
+    _ <- openDoc "src/OtherModule.hs" "haskell"
+    _ <- openDoc "src/OtherOtherModule.hs" "haskell"
+    _ <- openDoc "src/References.hs" "haskell"
+    _ <- openDoc "dependencyfoo/src/ModuleInDependency.hs" "haskell"
+    _ <- openDoc "dependencyfoo/src/OtherModuleInDependency.hs" "haskell"
+    liftIO $ sleep 2
+    f
```
Comment (peterwicksstringfield) [inline test/functional/Reference.hs:163]: This is a kludge to make sure references are ready before we ask for them.

**Comment by peterwicksstringfield** [comment]
8.8.3
https://hackage.haskell.org/package/ghc-8.8.3/docs/src/TyCoRep.html#Type
```
| FunTy Type Type     -- ^ t1 -> t2   Very common, so an important special case
```

vs

8.10.2
https://hackage.haskell.org/package/ghc-8.10.2/docs/src/TyCoRep.html#Type
```
  | FunTy      -- ^ t1 -> t2   Very common, so an important special case
                -- See Note [Function types]
     { ft_af  :: AnonArgFlag  -- Is this (->) or (=>)?
     , ft_arg :: Type           -- Argument type
     , ft_res :: Type }         -- Result type
```

**Pull request merged**